src/power_grid_model/core/power_grid_model.py

```diff
@@ -248,25 +259,54 @@ def _calculate(
             prepared_update.data_ptrs_per_component,
         )
 
-        # error handling
-        if not continue_on_batch_error:
-            assert_no_error(batch_size=batch_size)
-        else:
-            # continue on batch error
-            error: Optional[RuntimeError] = find_error(batch_size=batch_size)
-            if error is not None:
-                if isinstance(error, PowerGridBatchError):
-                    # continue on batch error
-                    self._batch_error = error
-                else:
-                    # raise normal error
-                    raise error
-
-        # flatten array for normal calculation
-        if not batch_calculation:
-            result_dict = {k: v.ravel() for k, v in result_dict.items()}
-
-        return result_dict
+        self._handle_errors(continue_on_batch_error=continue_on_batch_error, batch_size=batch_size)
+
+        return reduce_output_data(output_data=output_data, batch_calculation=batch_calculation)
+
+    # pylint: disable=too-many-arguments
+    def _calculate(
```

Comment:
I would remove this function, just call `_options` and `_calculate_impl` in power flow, state estimation, and short circuit.